### PR TITLE
[Routing] Fix localized paths

### PR DESCRIPTION
**Pull request by l-vo**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | 

Related to #40266, localized paths does not work anymore. This PR aims to fix that and add a rework on the tests (using real annotations/attributes instead of guessing the parsing that may lead to errors).

### CODE REVIEW

File: src/Symfony/Component/Routing/Tests/Fixtures/AnnotationFixtures/FooController.php
Comment (derrabus) [inline src/Symfony/Component/Routing/Tests/Fixtures/AnnotationFixtures/FooController.php]:
```suggestion
     * @Route(requirements={"locale":"en"})
```

There's no need to pass a path.

File: src/Symfony/Component/Routing/Tests/Fixtures/AnnotationFixtures/FooController.php
Comment (derrabus) [inline src/Symfony/Component/Routing/Tests/Fixtures/AnnotationFixtures/FooController.php]: Okay, right. This does work now that we've enabled `@NamedArgumentConstructor`, but I'd consider this to be an accident rather than intended behavior. I would not build a test case for that.

File: src/Symfony/Component/Routing/Annotation/Route.php
Comment (derrabus) [inline src/Symfony/Component/Routing/Annotation/Route.php]: This basically means that I could mix a localized path with other route properties:

```php
#[Route(['en' => '/readme', 'de' => '/liesmich', 'methods' => ['GET']])]
```

This wasn't possible before and we should not start to allow it now.

**Review by derrabus — CHANGES_REQUESTED** [review]
Good catch, thank you for your PR!

File: src/Symfony/Component/Routing/Annotation/Route.php
Comment (l-vo) [reply src/Symfony/Component/Routing/Annotation/Route.php]: Yes I did that because of the deprecation that must be triggered. But maybe I miss something about this deprecation. Which use case should trigger the deprecation "Passing an array as first argument to "%s" is deprecated" please ?

File: src/Symfony/Component/Routing/Tests/Fixtures/AnnotationFixtures/FooController.php
Comment (l-vo) [reply src/Symfony/Component/Routing/Tests/Fixtures/AnnotationFixtures/FooController.php]: related to https://github.com/symfony/symfony/pull/40767/files#r612721813. This fixture is to test deprecation but maybe I misunderstand the target of the deprecation.

Comment (derrabus) [reply src/Symfony/Component/Routing/Tests/Fixtures/AnnotationFixtures/FooController.php]: The deprecation is not supposed to be triggered by normal annotation usage. The target are manual constructor calls (that should not happen anyway).

File: src/Symfony/Component/Routing/Annotation/Route.php
Comment (derrabus) [reply src/Symfony/Component/Routing/Annotation/Route.php]: see my other comment

File: src/Symfony/Component/Routing/Tests/Fixtures/AnnotationFixtures/FooController.php
Comment (l-vo) [reply src/Symfony/Component/Routing/Tests/Fixtures/AnnotationFixtures/FooController.php]: Oh ok thank you. I pushed some changes :)

File: src/Symfony/Component/Routing/Annotation/Route.php
Comment (derrabus) [inline src/Symfony/Component/Routing/Annotation/Route.php]: Should be reverted.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Routing/Tests/Fixtures/AnnotationFixtures/FooController.php
Comment (l-vo) [reply src/Symfony/Component/Routing/Tests/Fixtures/AnnotationFixtures/FooController.php]: Sorry, I used cs fixer with a wrong PHP version;it broke attributes. I'll fix it tomorrow.

Comment (l-vo) [reply src/Symfony/Component/Routing/Tests/Fixtures/AnnotationFixtures/FooController.php]: Updated :)

**Comment by l-vo** [comment]
status: needs review

**Review by derrabus — APPROVED** [review]

**Comment by COil** [comment]
Hello, is there something to check related to this issue? I saw it wasn't shipped with 5.3 beta2.

**Comment by derrabus** [comment]
IIRC, we're only waiting for a second core team approval.

File: src/Symfony/Component/Routing/Annotation/Route.php
Comment (Tobion) [inline src/Symfony/Component/Routing/Annotation/Route.php:71]: @derrabus shouldn't the `$data` argument itself also be deprecated? Otherwise we cannot really remove the argument from the constructor. So $data should be removed from the list of arguments with the appropriate BC layer in place.

**Review by Tobion — APPROVED** [review]
The fix seems fine. But I think the deprecation in #40266 is not complete.

File: src/Symfony/Component/Routing/Annotation/Route.php
Comment (derrabus) [reply src/Symfony/Component/Routing/Annotation/Route.php:71]: We don't really have a strategy for phasing out named arguments, do we. I would consider the `$data` parameter deprecated already and my plan is to remove it from the constructor in Symfony 6.

I think we should about how proper deprecation layers for named arguments can look like in a dedicated RFC. And maybe we will fine-tune this deprecation layer for 5.4 then.

Comment (derrabus) [reply src/Symfony/Component/Routing/Annotation/Route.php:71]: I'm also not 100% happy with the current deprecation layer, I must admit.

**Comment by derrabus** [comment]
Thank you @l-vo.

**Pull request merged**
